File: Origin/reconstructed-project/src/combat/PikeAttackEffect.js
'use strict';

const { MeleeAttackEffect } = require('./MeleeAttackEffect');

// 来源: bundle.strings-decoded.js:24705-24749 / 命中段 24733-24741。
// 旋转 90ms、突刺 270ms 后由攻击效果开始结算，回收前再保留 120ms。
// PIKE_HIT_DELAY_MS / PIKE_EFFECT_DURATION_MS 为原始 Tween 链第三段 onStart 等价常量（bundle:24733-24741）：
//   段1 旋转 l/j_ → 段2 突刺 b[96]/j_（段末显隐 Qx 枪尖）→ 段3 回收 b[111]/j_（段首 onStart 调 U.LS() 命中）→ 段4 归位。
// 正式 Spine/Tween 动画事件接入后，此常量作「校准基线」与「无 Spine 回退」：
//   无 animationEventTimingProvider 注入时 hitAtMs = PIKE_HIT_DELAY_MS / playbackRate（与 v0.8.1 现状一致）；
//   正式动画事件到达时经 calibrateHitTiming 校准 hitAtMs 为事件时机，本常量为校准前的回退基线。
const PIKE_ATTACK_ROTATE_MS = 90;
const PIKE_ATTACK_THRUST_MS = 270;
const PIKE_HIT_DELAY_MS = PIKE_ATTACK_ROTATE_MS + PIKE_ATTACK_THRUST_MS;
const PIKE_EFFECT_DURATION_MS = PIKE_HIT_DELAY_MS + 120;

class PikeAttackEffect extends MeleeAttackEffect {
  constructor() { super('pike'); }

  reset() {
    // 枪兵专属字段重置：动画事件校准 provider、枪尖 Qx 表现 port 与视觉句柄。
    // 规则层命中字段（hitAtMs/elapsed/hitTriggered 等）由 super.reset() 处理。
    this.animationEventTimingProvider = null;
    this.pikeTipPresentation = null;
    this.pikeTipVisual = null;
    return super.reset();
  }

  launch({
    owner,
    target = null,
    enemyManager,
    damage = 0,
    radius = 48,
    playbackRate = 1,
    durationMs = null,
    // 可选：动画事件时机校准 provider（DEFERRED，正式 Spine/Tween 第三段 onStart 接入后由动画回调调 calibrateHitTiming）。
    // 无 provider 时命中回退 PIKE_HIT_DELAY_MS / playbackRate 常量基线（与 v0.8.1 现状一致）。
    animationEventTimingProvider = null,
    // 可选：枪尖 Qx 表现 port（DEFERRED 桩 no-op，实体 VFX 归 P2）。无 port 时跳过 Qx 视觉调度，不影响规则层命中。
    pikeTipPresentation = null,
  } = {}) {
    this.target = target;
    this.animationEventTimingProvider = animationEventTimingProvider || null;
    this.pikeTipPresentation = pikeTipPresentation || null;
    this.pikeTipVisual = null;
    const rate = Math.max(0.01, Number(playbackRate) || 1);
    // 命中时机基线：PIKE_HIT_DELAY_MS(360) / playbackRate（原始 Tween 链第三段 onStart 等价常量）。
    // 无论是否注入 provider，均以此常量基线起步——保证规则层始终有管理器驱动的非动画回退（满足 HANDOFF 行 440 约束）。
    // 正式动画事件到达时经 calibrateHitTiming 校准 hitAtMs 为事件时机，校准前此值为回退基线。
    const hitAtMs = PIKE_HIT_DELAY_MS / rate;
    const duration = durationMs == null
      ? PIKE_EFFECT_DURATION_MS / rate
      : Math.max(0, Number(durationMs) || 0);
    const result = super.launch({ owner, enemyManager, damage, radius, durationMs: duration, hitAtMs });

    // 4.4 枪尖 Qx 表现 port 调度（DEFERRED 桩 no-op，实体 VFX 归 P2）。
    // 对齐 bundle:24585/24736：突刺段（段2）创建 Qx 枪尖 pikeEff1.png 并 Tween.to("y",...) 位移。
    // port 为 DEFERRED 桩时 createPikeTipVisual 返回 null、animatePikeTipThrust no-op，不影响规则层命中结算。
    // 命中结算由 update()→hit() 驱动，不依赖 Qx 视觉对象。
    const presentation = this.pikeTipPresentation;
    if (presentation && typeof presentation.createPikeTipVisual === 'function' && this.owner) {
      const visual = presentation.createPikeTipVisual(this.owner);
      this.pikeTipVisual = visual; // 可能为 null（DEFERRED 桩），hidePikeTipVisual 容忍 null。
      if (visual && typeof presentation.animatePikeTipThrust === 'function') {
        // 突刺段时长 = PIKE_ATTACK_THRUST_MS(270) / playbackRate（bundle:24733-24741 段2 时长等价）。
        presentation.animatePikeTipThrust(visual, PIKE_ATTACK_THRUST_MS / rate);
      }
    }
    return result;
  }

  /**
   * 动画事件校准命中时机（覆盖基类 no-op 钩子，实现 2.3）。
   * 正式 Spine/Tween 接入后，动画第三段 onStart 到达时由动画回调（经 animationEventTimingProvider 路由）调用本方法，
   * 将 hitAtMs 校准为「当前 elapsed」（即动画事件到达时机），使下次 update() 满足 elapsed >= hitAtMs 触发 hit()。
   * 命中结算 MUST 仍走 update()→hit() 规则路径——本方法只重置 hitAtMs/elapsed 关系，不直接调 hit()（不倒退为动画回调直接结算）。
   * @param {number} _animationEventMs 动画事件到达时机（ms，相对效果启动）；当前校准目标为当前 elapsed（设计决策 1），
   *   参数保留供 P2 Spine 接入时按需记录事件时间戳，不影响校准语义。
   * @returns {boolean} true=已校准 hitAtMs（尚未命中且效果激活）；false=未校准（已命中/未激活，保持现状）。
   */
  calibrateHitTiming(_animationEventMs) {
    if (!this.active || this.hitTriggered) return false; // 已命中或未激活时不校准，避免重复结算。
    // 校准 hitAtMs 为当前 elapsed：下次 update() 即满足 elapsed >= hitAtMs 触发 hit()（规则路径）。
    this.hitAtMs = this.elapsed;
    return true;
  }

  cleanup(reason) {
    // 4.4 回收段枪尖 Qx 表现 port 调度（DEFERRED 桩 no-op）。
    // 对齐 bundle:24740：段3末 this.Qx.visible=false（枪尖特效回收段隐藏）。
    // 命中结算已由规则层 hit() 完成，不依赖本隐藏动作；port 为 DEFERRED 桩时 no-op 不影响规则层。
    const presentation = this.pikeTipPresentation;
    if (presentation && typeof presentation.hidePikeTipVisual === 'function' && this.pikeTipVisual) {
      presentation.hidePikeTipVisual(this.pikeTipVisual);
    }
    this.pikeTipVisual = null;
    this.pikeTipPresentation = null;
    this.animationEventTimingProvider = null;
    this.target = null;
    return super.cleanup(reason);
  }
}

module.exports = {
  PikeAttackEffect,
  PIKE_ATTACK_ROTATE_MS,
  PIKE_ATTACK_THRUST_MS,
  PIKE_HIT_DELAY_MS,
  PIKE_EFFECT_DURATION_MS,
};

File: UnityProject/Assets/GameScripts/HotFix/GameBattle/Combat/Effects/PikeAttackEffect.cs
using System.Collections.Generic;
using TEngine;

namespace GameBattle
{
    // ============================================================================
    // 任务 5.4：PikeAttackEffect —— 360ms 枪兵延迟命中效果
    // ----------------------------------------------------------------------------
    // 职责（design.md 第 1 节目录表 / Combat/Effects/PikeAttackEffect.cs）：
    //   以逻辑时间实现枪兵 360ms 命中。
    //
    // 来源证据（还原工程 PikeAttackEffect.js:1-111）：
    //   PikeAttackEffect 继承 MeleeAttackEffect，核心常量：
    //     PIKE_ATTACK_ROTATE_MS = 90      旋转段
    //     PIKE_ATTACK_THRUST_MS = 270     突刺段
    //     PIKE_HIT_DELAY_MS = 90+270 = 360   命中时机（旋转+突刺后）
    //     PIKE_EFFECT_DURATION_MS = 360+120 = 480   总持续时间（命中后保留 120ms 回收）
    //   launch({owner, target, enemyManager, damage, radius=48, playbackRate=1,
    //           durationMs=null, animationEventTimingProvider, pikeTipPresentation}):
    //     hitAtMs = PIKE_HIT_DELAY_MS / playbackRate   (=360 when rate=1)
    //     duration = durationMs ?? PIKE_EFFECT_DURATION_MS / playbackRate   (=480 when rate=1)
    //     super.launch({owner, enemyManager, damage, radius, durationMs:duration, hitAtMs})
    //   update/cleanup: 继承 MeleeAttackEffect（elapsed 累计、hit、duration-complete）
    //   calibrateHitTiming: 动画事件校准钩子（正式 Spine 接入用，当前 no-op 基线）
    //
    // 决策依据：
    //   - design.md 目录表："以逻辑时间实现枪兵 360ms 命中。"
    //   - task 5.4 约束：360ms PikeAttackEffect；实现 IAttackEffect +
    //     IPoolableBattleObject.ResetState；伤害经 AttackResolver；尊重 Settling/freeze；
    //     中文注释、UTF-8、LF。
    //   - C# 因 IAttackEffect 接口约束与 sealed 设计，PikeAttackEffect 不继承
    //     MeleeAttackEffect（JS 继承），改为独立 sealed class 复用 MeleeAttackEffect 的
    //     命中/时序逻辑蓝本。共享逻辑（hit 查询+去重+resolver 提交）在本类型内独立实现。
    //   - task 5.4 playbackRate：task 指定 360ms 固定值，playbackRate 默认 1.0，
    //     本类型固定 hitAtMs=360/durationMs=480，不支持变速（与最简移植一致）。
    //
    // 不变量：
    //   1. 唯一推进：Update 只由 AttackEffectManager.Update 调用。
    //   2. 360ms 命中：elapsed >= 360ms 时 Hit（范围查询+去重+resolver 提交）。
    //   3. 480ms 完成：elapsed >= 480ms 时标记非活动（命中后保留 120ms 回收段）。
    //   4. 伤害经 resolver：Hit 调 resolver.QueryEnemyObjects + resolver.Hit。
    //   5. hitSet 去重：同效果内同一敌人只命中一次。
    //   6. Cancel 不伤害；ResetState 完整清空；幂等。
    // ============================================================================

    /// <summary>
    /// 枪兵 360ms 延迟命中效果：以逻辑时间实现命中时序（task 5.4）。
    /// </summary>
    /// <remarks>
    /// <para><b>职责（design.md 目录表）：</b>以逻辑时间实现枪兵 360ms 命中。
    /// 替代还原工程 <c>PikeAttackEffect.js</c>（继承 MeleeAttackEffect，C# 改为独立 sealed）。</para>
    ///
    /// <para><b>时序常量（对应 JS <c>PIKE_*_MS</c>）：</b>
    /// <list type="bullet">
    /// <item><see cref="PikeAttackRotateMs"/>=90：旋转段（bundle 段1）</item>
    /// <item><see cref="PikeAttackThrustMs"/>=270：突刺段（bundle 段2）</item>
    /// <item><see cref="PikeHitDelayMs"/>=360：命中时机=旋转+突刺（段3 onStart 等价常量）</item>
    /// <item><see cref="PikeEffectDurationMs"/>=480：总持续时间=命中+120ms 回收段</item>
    /// </list></para>
    ///
    /// <para><b>360ms 命中（task 5.4 "360ms PikeAttackEffect"）：</b>
    /// <see cref="Update"/> 累计 <c>elapsed</c>，<c>elapsed &gt;= 360ms</c> 时触发
    /// <see cref="Hit"/>（范围查询+去重+resolver 提交）。<c>elapsed &gt;= 480ms</c> 时
    /// 标记非活动完成（命中后保留 120ms 回收段，对应 JS <c>PIKE_EFFECT_DURATION_MS</c>）。</para>
    ///
    /// <para><b>伤害经 AttackResolver（task 5.4 约束）：</b>
    /// <see cref="Hit"/> 经 <see cref="AttackResolver.QueryEnemyObjects"/> 查询、
    /// <see cref="AttackResolver.Hit"/> 提交，不直接 <see cref="IEnemyEntity.Hit"/>。</para>
    ///
    /// <para><b>hitSet 去重：</b>同效果内同一敌人只命中一次（对应 JS 继承的 hitSet）。</para>
    ///
    /// <para><b>Cancel 不伤害（spec "Settling has no gameplay damage authority"）。</b></para>
    ///
    /// <para><b>本类型为 internal sealed：</b>只供 GameBattle 内部使用。</para>
    /// </remarks>
    internal sealed class PikeAttackEffect : IAttackEffect
    {
        // ====================================================================
        // 时序常量（对应 JS PIKE_*_MS，bundle.strings-decoded.js:24705-24749）
        // ====================================================================

        /// <summary>枪兵攻击旋转段时长（对应 <c>PIKE_ATTACK_ROTATE_MS=90</c>，bundle 段1）。</summary>
        private const long PikeAttackRotateMs = 90;

        /// <summary>枪兵攻击突刺段时长（对应 <c>PIKE_ATTACK_THRUST_MS=270</c>，bundle 段2）。</summary>
        private const long PikeAttackThrustMs = 270;

        /// <summary>枪兵命中延迟=旋转+突刺（对应 <c>PIKE_HIT_DELAY_MS=360</c>，段3 onStart 等价常量）。</summary>
        private const long PikeHitDelayMs = PikeAttackRotateMs + PikeAttackThrustMs; // 360

        /// <summary>枪兵效果总持续时间=命中延迟+120ms 回收段（对应 <c>PIKE_EFFECT_DURATION_MS=480</c>）。</summary>
        private const long PikeEffectDurationMs = PikeHitDelayMs + 120; // 480

        // ====================================================================
        // 日志标签
        // ====================================================================

        /// <summary>日志标签前缀。</summary>
        private const string LogTag = "[PikeAttackEffect]";

        // ====================================================================
        // 活动状态与所有者
        // ====================================================================

        /// <summary>是否活动。</summary>
        private bool _active;

        /// <summary>所有者引用。</summary>
        private IAttackEffectOwner _owner;

        // ====================================================================
        // 外部依赖
        // ====================================================================

        /// <summary>攻击解析服务。</summary>
        private AttackResolver _resolver;

        /// <summary>敌人管理器。</summary>
        private EnemyManager _enemyManager;

        // ====================================================================
        // 攻击参数
        // ====================================================================

        /// <summary>基础伤害值。</summary>
        private int _damage;

        /// <summary>命中半径（默认 48，对应 JS <c>radius=48</c>）。</summary>
        private float _radius;

        /// <summary>敌人格子宽。</summary>
        private float _cellWidth;

        /// <summary>敌人格子高。</summary>
        private float _cellHeight;

        // ====================================================================
        // 时序状态
        // ====================================================================

        /// <summary>已累计时长（毫秒）。</summary>
        private long _elapsed;

        /// <summary>是否已命中，防止重复命中。</summary>
        private bool _hitTriggered;

        // ====================================================================
        // 已命中敌人 ID 集合（对应 JS 继承的 hitSet）
        // ====================================================================

        /// <summary>已命中敌人 ID 集合，同效果内同一敌人只命中一次。</summary>
        private readonly HashSet<int> _hitSet = new HashSet<int>();

        // ====================================================================
        // IAttackEffect 属性
        // ====================================================================

        /// <inheritdoc />
        public bool Active => _active;

        /// <inheritdoc />
        public object Owner => _owner;

        // ====================================================================
        // 构造
        // ====================================================================

        /// <summary>构造枪兵攻击效果。池工厂调用。</summary>
        internal PikeAttackEffect()
        {
            ResetState();
        }

        // ====================================================================
        // Launch —— 由单位在 Acquire 后调用
        // --------------------------------------------------------------------
        // 对应 JS PikeAttackEffect.launch（简化：去掉 playbackRate/动画校准/枪尖表现 port）。
        // hitAtMs 固定 360，durationMs 固定 480（task 5.4 最简移植）。
        // ====================================================================

        /// <summary>
        /// 启动枪兵延迟命中效果：配置参数并激活。
        /// </summary>
        /// <param name="owner">所有者（非 null）。</param>
        /// <param name="resolver">攻击解析服务（非 null）。</param>
        /// <param name="enemyManager">敌人管理器（非 null）。</param>
        /// <param name="damage">基础伤害值（正数）。</param>
        /// <param name="cellWidth">敌人格子宽。</param>
        /// <param name="cellHeight">敌人格子高。</param>
        /// <param name="radius">命中半径（默认 48）。</param>
        /// <returns>this。</returns>
        /// <remarks>
        /// <para><b>对应 JS <c>launch</c>：</b>固定 <c>hitAtMs=PikeHitDelayMs(360)</c>、
        /// <c>durationMs=PikeEffectDurationMs(480)</c>。JS 按 playbackRate 缩放，C# 按 task 5.4
        /// 固定 360/480（playbackRate=1.0）。动画事件校准 provider 与枪尖 Qx 表现 port
        /// （JS DEFERRED 桩）本期不移植——规则层命中由 update→hit 驱动，不依赖视觉对象。</para>
        /// </remarks>
        internal PikeAttackEffect Launch(
            IAttackEffectOwner owner,
            AttackResolver resolver,
            EnemyManager enemyManager,
            int damage,
            float cellWidth,
            float cellHeight,
            float radius = 48f)
        {
            _owner = owner;
            _resolver = resolver;
            _enemyManager = enemyManager;
            _damage = damage;
            _radius = radius;
            _cellWidth = cellWidth;
            _cellHeight = cellHeight;
            _elapsed = 0;
            _hitTriggered = false;
            _hitSet.Clear();
            _active = true;
            return this;
        }

        // ====================================================================
        // Update —— 唯一推进入口
        // ====================================================================

        /// <inheritdoc />
        /// <summary>
        /// 推进一帧：累计 elapsed，到达 360ms 触发命中，到达 480ms 标记非活动完成。
        /// </summary>
        /// <param name="deltaMs">子步时长（毫秒）。</param>
        /// <remarks>
        /// <para><b>对应 JS 继承的 <c>update</c>：</b>
        /// elapsed += delta; elapsed&gt;=hitAtMs(360) 且未命中 → hit();
        /// elapsed&gt;=durationMs(480) → cleanup（C# 置 active=false）。</para>
        /// <para><b>命中作为同步副作用</b>，伤害立即生效。</para>
        /// </remarks>
        public void Update(long deltaMs)
        {
            if (!_active)
            {
                return;
            }

            if (deltaMs > 0)
            {
                _elapsed += deltaMs;
            }

            // 命中时机：elapsed >= 360ms → 触发命中（对应 JS hitAtMs=PIKE_HIT_DELAY_MS）。
            if (!_hitTriggered && _elapsed >= PikeHitDelayMs)
            {
                _hitTriggered = true;
                Hit();
            }

            // 完成判定：elapsed >= 480ms → 标记非活动（对应 JS durationMs=PIKE_EFFECT_DURATION_MS）。
            if (_elapsed >= PikeEffectDurationMs)
            {
                _active = false;
            }
        }

        // ====================================================================
        // Hit —— 经 AttackResolver 查询目标并提交伤害（复用 MeleeAttackEffect 蓝本）
        // ====================================================================

        /// <summary>
        /// 命中结算：查询半径内敌人，经 <see cref="AttackResolver"/> 提交伤害。
        /// </summary>
        /// <remarks>
        /// <para>逻辑与 <see cref="MeleeAttackEffect"/> 的 Hit 一致（JS 中 PikeAttackEffect
        /// 继承 MeleeAttackEffect.hit）。查询 center 周围 radius 内可攻击目标，hitSet 去重，
        /// resolver.Hit 提交。枪兵无 multiplier（JS 默认 1）。</para>
        /// </remarks>
        private void Hit()
        {
            if (!_active || _owner == null || _enemyManager == null || _resolver == null)
            {
                return;
            }

            List<IEnemyEntity> targets = _resolver.QueryEnemyObjects(
                _enemyManager,
                _owner.CenterX, _owner.CenterY,
                _radius,
                _owner.Side,
                _cellWidth, _cellHeight,
                null);

            if (targets == null || targets.Count == 0)
            {
                return;
            }

            // 枪兵无 multiplier（JS super.launch 未传 multiplier，默认 1）。
            int count = targets.Count;
            for (int i = 0; i < count; i++)
            {
                IEnemyEntity target = targets[i];
                if (target == null)
                {
                    continue;
                }

                if (_hitSet.Contains(target.Id))
                {
                    continue;
                }

                _hitSet.Add(target.Id);
                _resolver.Hit(target, _damage, _owner.RuntimeId);
            }
        }

        // ====================================================================
        // Cancel —— 取消并清理（不造成伤害）
        // ====================================================================

        /// <inheritdoc />
        public void Cancel(string reason)
        {
            if (!_active)
            {
                return;
            }

            _active = false;
            _owner = null;
            _enemyManager = null;
            _resolver = null;
            _hitSet.Clear();
        }

        // ====================================================================
        // ResetState —— 池回收时清空全部可变状态
        // ====================================================================

        /// <inheritdoc />
        public void ResetState()
        {
            _active = false;
            _owner = null;
            _resolver = null;
            _enemyManager = null;
            _damage = 0;
            _radius = 0f;
            _cellWidth = 0f;
            _cellHeight = 0f;
            _elapsed = 0;
            _hitTriggered = false;
            _hitSet.Clear();
        }
    }
}
